apps/common/src/main/scala/org/lfdecentralizedtrust/splice/store/db/DbMultiDomainAcsStore.scala

```diff
@@ -229,11 +229,31 @@ final class DbMultiDomainAcsStore[TXE](
       .value
   }
 
-  def hasArchived(ids: Seq[ContractId[?]])(implicit
-      traceContext: TraceContext
-  ): Future[Boolean] =
-    // TODO(#838): implement this as a single DB query
-    Future.sequence(ids.map(lookupContractStateById)).map(_.exists(_.isEmpty))
+  def containsArchived(ids: Seq[ContractId[?]])(implicit
+                                                traceContext: TraceContext
+  ): Future[Boolean] = waitUntilAcsIngested {
+    if (ids.isEmpty) Future.successful(true)
+    else {
+      val contractIds = inClause(ids)
+      val expectedCount = ids.size
+      storage.query(
+        (sql"""
+        select count(1) from (
+           select acs.contract_id
+           from #$acsTableName acs
+           where acs.store_id = $acsStoreId
+           and acs.migration_id = $domainMigrationId
+           and acs.contract_id in """ ++ contractIds ++ sql"""
```

Comment:
this is not gonna work for very large numbers of contract ids. Have you double checked that the use sites are ok with this?

Reply:
In the use I've seen we were limiting it to 20 now. But we can add a safety check that if we have really high batches then we split them.